### PR TITLE
Implement manual particle emission and particle sub emitters.

### PR DESCRIPTION
**Pull request by reduz**
Uses compute shader atomics to implement manual particle emission and sub emitters.
sub_emitter

### CODE REVIEW

**Comment by bruvzg** [comment]
Shader compilation fails on MoltenVK:
```
ERROR:  - Message Id Number: 0 | Message Id Name:
	VK_ERROR_INITIALIZATION_FAILED: Shader library compile failed (Error code 3):
Compilation failed:

program_source:138:92: error: no viable overloaded operator[] for type 'volatile device metal::float4x4' (aka 'volatile device matrix<float, 4, 4>')
                    particles.data[particle].xform[3] = src_particles.data[src_index].xform[3];
                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
program_source:147:92: error: no viable overloaded operator[] for type 'volatile device metal::float4x4' (aka 'volatile device matrix<float, 4, 4>')
                    particles.data[particle].xform[0] = src_particles.data[src_index].xform[0];
                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
program_source:148:92: error: no viable overloaded operator[] for type 'volatile device metal::float4x4' (aka 'volatile device matrix<float, 4, 4>')
                    particles.data[particle].xform[1] = src_particles.data[src_index].xform[1];
                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
program_source:149:92: error: no viable overloaded operator[] for type 'volatile device metal::float4x4' (aka 'volatile device matrix<float, 4, 4>')
                    particles.data[particle].xform[2] = src_particles.data[src_index].xform[2];
                                                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~
```

**Comment by reduz** [comment]
@bruvzg I thought empty arrays were allowed nowadays in C++, being 2020 but seems its not the case, so had to change the size to 1.

**Comment by bruvzg** [comment]
> I thought empty arrays were allowed nowadays in C++, being 2020 but seems its not the case, so had to change the size to 1.

It's shader build error not engine build, zero size arrays do trigger warning but it's still building whiteout `werror=yes` (there's also invalid type `uint pad` that I have changed for my test build).

**Comment by akien-mga** [comment]
> (there's also invalid type uint pad that I have changed for my test build)

Indeed:
```
D:\a\godot\godot\servers/rendering/rasterizer_rd/rasterizer_storage_rd.h(607): error C3646: 'pad': unknown override specifier
D:\a\godot\godot\servers/rendering/rasterizer_rd/rasterizer_storage_rd.h(607): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
```

**Comment by reduz** [comment]
ok lets hope it builds now

**Comment by bruvzg** [comment]
> ok lets hope it builds now

It builds with `werror=yes` now, but the same shader error is still there.

File: servers/rendering/rasterizer_rd/shaders/particles.glsl [hunk 1]
```diff
@@ -66,6 +66,38 @@ layout(set = 1, binding = 1, std430) restrict buffer Particles {
...
+
+struct ParticleEmission {
+	mat4 xform;
+	vec3 velocity;
+	uint flags;
+	vec4 color;
+	vec4 custom;
+};
+
+layout(set = 1, binding = 2, std430) restrict volatile coherent buffer SourceEmission {
```
Comment (bruvzg) [inline servers/rendering/rasterizer_rd/shaders/particles.glsl:83]: Removing `volatile` and `coherent` qualifiers from here fixes shader compilation error, not sure what consequences this has. `GPUParticles3D` seems to be broken regardless of this change.

```suggestion
layout(set = 1, binding = 2, std430) restrict buffer SourceEmission {
```

**Pull request merged**